OCaml_Intro_to_CS/part_6.ml

```diff
@@ -0,0 +1,220 @@
+(*
+Suppose you have two functions f : ‘a -> ‘b and g :
+‘c -> ‘a. Define their functional composition as follows:
+  *)
+
+let compose f g = fun x -> f (g x);;
+
+let inc x = x + 1;;
+
+let double x = x + x;;
+
+let h = compose inc double;;
+
+let result = h 5 ;;
+
+let () =
+  Printf.printf "result of compose inc double 5: %d\n" result;; 
+
+(*
+Please implement the compose_list function. Note: if the list is empty,
+then please return the argument (that is, be an identity function). In other words:
+# compose_list [] 7;;
+- : int = 7
+*)
+
+let dec x = x - 1;;
+
+let compose_list lst x =
+  let rec aux acc = function
+    | [] -> if acc = 0  then 
+      (
+        print_endline "Reached Identity";
+        x (* identity *)
+      )
+      else (
+        print_endline "Reached end of list without identity";
+        acc 
+      )
+    | h :: t -> 
+        if acc = 0 then 
+          (
+            print_endline "Applying function from list";
+            aux (h x) t 
+          )
+        else 
+          (
+            print_endline "Applying function from list with acc";
+            aux (h acc) t
+          )
+  in
+  aux 0 lst
+;;
```

Comment:
Yeapp because im returning the acc given the condition `[]`, so in that case, the return value has to be an int since i defined my acc as 0.

but if we're passing the value given in the function `x` instead, it could just be any type `a since we could pass a string, and have each function manipulate the string, and when we reach the `[]` it will return whatever end result that the value `x` has been called through the function list.

Is this accurate?

Reply:
Yes that's right, the type is inferred as int because of the accumulator

Reply:
and you return your accumulator on the empty list

Reply:
You could also drop the accumulator entirely, since it's available in the outer scope
```ocaml
let compose_lst lst x =
  let rec aux = function
  | [] -> x
  | h ::  t -> h (aux t)
```
but what you have now is correct